cmd/ceremony/main.go

```diff
@@ -0,0 +1,280 @@
+// ceremony is a key/certificate ceremony tool.
+package main
+
+import (
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+
+	"github.com/letsencrypt/boulder/pkcs11helpers"
+)
+
+type ceremonyType int
+
+const (
+	rootCeremony ceremonyType = iota
+	intermediateCeremony
+	keyCeremony
+)
+
+type ceremonyConfig struct {
+	PKCS11Module       string       `yaml:"pkcs11-module"`
+	CeremonyType       string       `yaml:"ceremony-type"`
+	KeySlot            uint         `yaml:"key-slot"`
+	KeyLabel           string       `yaml:"key-label"`
+	KeyID              string       `yaml:"key-id"`
+	KeyType            string       `yaml:"key-type"`
+	ECDSACurve         string       `yaml:"ecdsa-curve"`
+	PublicKeyPath      string       `yaml:"public-key-path"`
+	CertificatePath    string       `yaml:"certificate-path"`
+	IssuerPath         string       `yaml:"issuer-path"`
+	CertificateProfile *certProfile `yaml:"certificate-profile"`
+}
+
+func (cc ceremonyConfig) Validate() error {
+	if cc.PKCS11Module == "" {
+		return errors.New("pkcs11-module is required")
+	}
+	if cc.CeremonyType != "root" && cc.CeremonyType != "intermediate" && cc.CeremonyType != "key" {
+		return errors.New("ceremony-type can only be 'root', 'intermediate', or 'key'")
+	}
+	switch cc.CeremonyType {
+	case "root":
+		if cc.KeyLabel == "" {
```

Comment:
Some of these checks, like KeyLabel and KeyID, are redundant across all three ceremony types. Could you factor those out so they are checked in all three?

Also, please split out what remains for each case into its own `func`.